// jquery-file-upload-middleware/lib/uploadhandler.js
var EventEmitter = require('events').EventEmitter,
    path = require('path'),
    fs = require('fs'),
    easyimg = require('easyimage'),
    formidable = require('formidable'),
    imageMagick = require('imagemagick'),
    mkdirp = require('mkdirp'),
    _ = require('lodash');

module.exports = function (options) {

    var FileInfo = require('./fileinfo')(
        _.extend({
            baseDir: options.uploadDir
        }, _.pick(options, 'minFileSize', 'maxFileSize', 'acceptFileTypes'))
    );

    var UploadHandler = function (req, res, callback) {
        EventEmitter.call(this);
        this.req = req;
        this.res = res;
        this.callback = callback;
    };
    require('util').inherits(UploadHandler, EventEmitter);

    UploadHandler.prototype.noCache = function () {
        this.res.set({
            'Pragma': 'no-cache',
            'Cache-Control': 'no-store, no-cache, must-revalidate',
            'Content-Disposition': 'inline; filename="files.json"'
        });
    };

    UploadHandler.prototype.get = function () {
        this.noCache();
        var files = [];
        fs.readdir(options.uploadDir(), _.bind(function (err, list) {
            _.each(list, function (name) {
                var stats = fs.statSync(options.uploadDir() + '/' + name),
                    fileInfo;
                if (stats.isFile()) {
                    fileInfo = new FileInfo({
                        name: name,
                        size: stats.size
                    });
                    this.initUrls(fileInfo);
                    files.push(fileInfo);
                }
            }, this);
            this.callback(files);
        }, this));
    };

    UploadHandler.prototype.post = function () {

        var self = this,
            form = new formidable.IncomingForm(),
            tmpFiles = [],
            files = [],
            map = {},
            counter = 1,
            redirect,
            finish = _.bind(function () {
                if (!--counter) {
                    _.each(files, function (fileInfo) {
                        this.initUrls(fileInfo);
                        this.emit('end', fileInfo);
                    }, this);
                    this.callback(files, redirect);
                }
            }, this);

        this.noCache();

        form.uploadDir = options.tmpDir;
        form
            .on('fileBegin', function (name, file) {
                tmpFiles.push(file.path);
                var fileInfo = new FileInfo(file);
                fileInfo.safeName();
                map[path.basename(file.path)] = fileInfo;
                files.push(fileInfo);
                self.emit('begin', fileInfo);

            })
            .on('field', function (name, value) {
                if (name === 'redirect') {
                    redirect = value;
                }
            })
            .on('file', function (name, file) {
                var fileInfo = map[path.basename(file.path)];
                if (fs.existsSync(file.path)) {
                    fileInfo.size = file.size;
                    if (!fileInfo.validate()) {
                        fs.unlink(file.path);
                        return;
                    }

                    //new convert image
                    var convertImage = function (callback) {


                            _.each(options.imageVersions, function (value, version) {
                                // creating directory recursive
                                if (!fs.existsSync(options.jpegDir + '/'))
                                    mkdirp.sync(options.jpegDir + '/');

                                counter++;
                                var opts = options.imageVersions[version];
                                var quality;
                                if(opts.quality == 'auto'){
                                    var stats = fs.lstatSync(options.uploadDir() + '/' + fileInfo.name);
                                    if(stats.size > 100000000) quality = 20;
                                    else if(stats.size > 50000000) quality = 30;
                                    else if(stats.size > 40000000) quality = 40;
                                    else if(stats.size > 30000000) quality = 50;
                                    else if(stats.size > 20000000) quality = 60;
                                    else if(stats.size > 10000000) quality = 70;
                                    else if(stats.size > 5000000) quality = 80;
                                    else quality = 90;
                                }
                                else{
                                    quality = opts.quality;
                                }
                                console.log(quality);
                                easyimg.convert({src:options.uploadDir() + '/' + fileInfo.name, dst:options.jpegDir + '/' + path.basename(fileInfo.name, path.extname(fileInfo.name)) + '.jpeg', quality:quality}, function(err, stdout, stderr) {
                                    if (err) throw err;
                                    console.log('Converted to jpeg');
                                    generatePreviews();
                                    finish();
                                });


                                // creating directory recursive
                                if (!fs.existsSync(options.tiffDir + '/'))
                                    mkdirp.sync(options.tiffDir + '/');

                                counter++;
                                easyimg.convert({src:options.uploadDir() + '/' + fileInfo.name, dst:options.tiffDir + '/' + path.basename(fileInfo.name, path.extname(fileInfo.name)) + '.tiff'}, function(err, stdout, stderr) {
                                    if (err) throw err;
                                    console.log('Converted to tiff');
                                    finish();
                                });
                            });

                    }


                    var generatePreviews = function () {

                            _.each(options.imageVersions, function (value, version) {
                                // creating directory recursive
                                if (!fs.existsSync(options.thumbnailDir + '/'))
                                    mkdirp.sync(options.thumbnailDir + '/');

                                counter++;
                                var opts = options.imageVersions[version];
                                imageMagick.resize({
                                    width: opts.width,
                                    height: opts.height,
                                    srcPath: options.jpegDir + '/' + path.basename(fileInfo.name, path.extname(fileInfo.name)) + '.jpeg',
                                    dstPath: options.thumbnailDir + '/' + path.basename(fileInfo.name, path.extname(fileInfo.name)) + '.jpeg',
                                    customArgs: opts.imageArgs || ['-auto-orient']
                                }, finish);
                            });

                    }

                    //---------------

                    if (!fs.existsSync(options.uploadDir() + '/'))
                        mkdirp.sync(options.uploadDir() + '/');

                    counter++;
                    fs.rename(file.path, options.uploadDir() + '/' + fileInfo.name, function (err) {
                        if (!err) {
                            convertImage();
                            finish();
                        } else {

                            var is = fs.createReadStream(file.path);
                            var os = fs.createWriteStream(options.uploadDir() + '/' + fileInfo.name);
                            is.on('end', function (err) {
                                if (!err) {
                                    fs.unlinkSync(file.path);
                                    convertImage();
                                }
                                finish();
                            });
                            is.pipe(os);
                        }
                    });
                }
            })
            .on('aborted', function () {
                _.each(tmpFiles, function (file) {
                    var fileInfo = map[path.basename(file)];
                    self.emit('abort', fileInfo);
                    fs.unlink(file);
                });
            })
            .on('error', function (e) {
                self.emit('error', e);
            })
            .on('progress', function (bytesReceived, bytesExpected) {
                if (bytesReceived > options.maxPostSize)
                    self.req.connection.destroy();
            })
            .on('end', finish)
            .parse(self.req);
    };

    UploadHandler.prototype.destroy = function () {
        var self = this,
            fileName = path.basename(decodeURIComponent(this.req.url));

        fs.unlink(options.uploadDir() + '/' + fileName, function (ex) {

            //added
            fs.unlink(options.jpegDir + '/' + path.basename(fileName, path.extname(fileName)) + '.jpeg');
            fs.unlink(options.tiffDir + '/' + path.basename(fileName, path.extname(fileName)) + '.tiff');
            //----------
            _.each(options.imageVersions, function (value, version) {
                fs.unlink(options.thumbnailDir + '/' + path.basename(fileName, path.extname(fileName)) + '.jpeg');
            });
            self.emit('delete', fileName);
            self.callback(!ex);
        });
    };

    UploadHandler.prototype.initUrls = function (fileInfo) {
        var baseUrl = (options.ssl ? 'https:' : 'http:') + '//' + (options.hostname || this.req.get('Host'));
        fileInfo.setUrl(null, baseUrl + options.uploadUrl() + '/' + path.basename(options.uploadDir()));
        fileInfo.setUrl('delete', baseUrl + this.req.originalUrl);
        _.each(options.imageVersions, function (value, version) {
            if (fs.existsSync(options.thumbnailDir + '/' + path.basename(fileInfo.name, path.extname(fileInfo.name)) + '.jpeg')) {
                fileInfo.setUrl(version, baseUrl + options.uploadUrl() + '/' + version);
            }
        }, this);
    };

    return UploadHandler;
}
